src/Shared/WeakStringCache.cs

```diff
@@ -0,0 +1,355 @@
+                    WeakHandleOverflow = new List<GCHandle>(2);
+                    WeakHandleOverflow.Add(GCHandle.Alloc(str, GCHandleType.Weak));
+                }
+                else
+                {
+                    // Find the first usable slot in the overflow area or add a new slot if there is none.
+                    bool foundExistingSlot = false;
+                    for (int i = 0; i < WeakHandleOverflow.Count; i++)
+                    {
+                        if (WeakHandleOverflow[i].Target == null)
+                        {
+                            GCHandle handle = WeakHandleOverflow[i];
+                            handle.Target = str;
+                            WeakHandleOverflow[i] = handle;
+                            foundExistingSlot = true;
+                            break;
+                        }
+                    }
+                    if (!foundExistingSlot)
+                    {
+                        WeakHandleOverflow.Add(GCHandle.Alloc(str, GCHandleType.Weak));
+                    }
+                }
+            }
+
+            /// <summary>
+            /// Frees all GC handles allocated in this bucket.
+            /// </summary>
+            public void Free()
+            {
+                WeakHandle.Free();
+                if (WeakHandleOverflow != null)
+                {
+                    for (int i = 0; i < WeakHandleOverflow.Count; i++)
+                    {
+                        WeakHandleOverflow[i].Free();
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Improvised capacity for the scavenging heuristics.
+        /// </summary>
+        private int _capacity = 100;
+
+        /// <summary>
+        /// A dictionary mapping string hash codes to string buckets holding the actual weak GC handles.
+        /// </summary>
+        private Dictionary<int, StringBucket> _stringsByHashCode;
+
+        public WeakStringCache()
+        {
+            _stringsByHashCode = new Dictionary<int, StringBucket>(_capacity);
+        }
+
+        /// <summary>
+        /// Main entrypoint of this cache. Tries to look up a string that matches the given internable. If it succeeds, returns
+        /// the string and sets cacheHit to true. If the string is not found, calls ExpensiveConvertToString on the internable,
+        /// adds the resulting string to the cache, and returns it setting cacheHit to false.
+        /// </summary>
+        /// <param name="internable">The internable describing the string we're looking for.</param>
+        /// <returns>A string matching the given internable.</returns>
+        public string GetOrCreateEntry<T>(T internable, out bool cacheHit) where T : IInternable
+        {
+            int hashCode = GetInternableHashCode(internable);
+
+            StringBucket bucket;
+            string result;
+            lock (_stringsByHashCode)
```

Comment:
Would an upgradeable read/write lock be a good fit here?

Reply:
The workload is mostly single-threaded and `ReaderWriterLockSlim` has almost twice the overhead if there's no contention compared to a simple lock. So not a good fit in my opinion as the benefits of R/W lock will likely not materialize here.